### PR TITLE
setting sync_dist to true for validation metrics

### PR DESCRIPTION
**Pull request by krishnacpuvvada**
# What does this PR do ?

setting sync_dist=True for `on_validation_epoch_end` in `ModelPT.py` 
This is in response to the recent slack discussion on "wiping out of checkpoints mid training and starting from scratch" in Canary training.

**Collection**: [all] - The change is in ModelPT.py

# Changelog 
- setting sync_dist=true in `on_validation_epoch_end`

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [X] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

**Review by titu1994 — APPROVED** [review]
This can cause hangs, but for now let's try it.

**Comment by pzelasko** [comment]
This is on validation epoch end so only a single sync per validation loop. That should be OK both hang-wise and speed-wise.

**Comment by titu1994** [comment]
Makes sense

**Review by pzelasko — APPROVED** [review]

File: nemo/core/classes/modelPT.py [hunk 1]
```diff
@@ -1025,7 +1025,7 @@ def on_validation_epoch_end(self) -> Optional[Dict[str, Dict[str, torch.Tensor]]
                 self.validation_step_outputs[dataloader_idx].clear()  # free memory
 
             if 'log' in output_dict:
-                self.log_dict(output_dict.pop('log'), on_epoch=True)
+                self.log_dict(output_dict.pop('log'), on_epoch=True, sync_dist=True, reduce_fx='mean')
```
Comment (maanug-nv) [inline nemo/core/classes/modelPT.py:1028]: Just FYI, `reduce_fx` doesn't apply to `sync_dist`, it applies to `on_epoch`, ie reduction over step values ([docs](https://lightning.ai/docs/pytorch/stable/api/lightning.pytorch.core.LightningModule.html#lightning.pytorch.core.LightningModule.log). 
Since this change is in `on_validation_epoch_end()`, I'm not sure if the `reduce_fx='mean'` will do anything, but that's the default anyways.

**Comment by maanug-nv** [comment]
Another note, from my experience, `sync_dist=True` is tricky with parallelisms, especially pipeline parallelisms. If some metrics are non-existent on some ranks (eg val_loss), this could cause a hang. If the metric is 0.0 on some ranks, the mean reduction over ranks could be wrong or not produce intended value. 

@pzelasko  I know there's a slack thread on this issue with more context, but I was hoping you could summarize the issue and why this will fix. I've been fixing some logging bugs in NeMo 2.0 recently, so I'd just like to give my input if there's similarity.

File: nemo/core/classes/modelPT.py (hunk 1 above)
Comment (maanug-nv) [reply nemo/core/classes/modelPT.py:1028]: Also FYI, the reduction function for syncing over ranks is set [here](https://github.com/Lightning-AI/pytorch-lightning/blob/master/src/lightning/pytorch/core/module.py#L539)

**Comment by pzelasko** [comment]
@krishnacpuvvada @titu1994 what's the verdict on this one? Do we need a Canary-specific mechanism to synchronize the validation metrics if it can't be done at the modelPT level?

**Comment by titu1994** [comment]
(1) Don't push this to modelPT. 
(2) If Canary needs it, it can override this function specifically and copy paste the code (for now). 

Do note - @nithinraok plans to remove modelpt level support for multi validation and test data loader, which basically would mean every model has to handle this stuff manually anyway.
